Columns - fix margins, keep 2-col breakpoint

drdogbot7 opened the pull request:
This is a revision of #12408 which includes better comments and also applies the same fixes in the editor.

## Background
There should be 3 breakpoints:

**Mobile: less than 600px**
all columns full width and stacked. No margins.

**Small: 600px - 782px**
columns wrap in rows of 2. Margins on the inside.

**Medium: bigger than 782px**
all columns fit in one row. Margins on the inside.

## PROBLEMS!

**PROBLEM 1**
Columns don't wrap properly at Small breakpoint

![image](https://user-images.githubusercontent.com/1050936/49835575-3f47e280-fd65-11e8-975e-41e590392cbc.png)

**SOLUTION**
- move `not` selectors to Medium breakpoint
- adjust flex-basis to account for margins using calc()


**PROBLEM 2**
At the Medium breakpoint, when there is an odd number of columns, and extra margin is added to the last item.

![image](https://user-images.githubusercontent.com/1050936/49835697-a82f5a80-fd65-11e8-9ee2-504f1a12b68b.png)

**SOLUTION**
- reset `odd` and `even` selectors to `initial` at the Medium breakpoint

## How has this been tested?
tested in local wordpress docker env

ORIG:
https://codepen.io/drdogbot7/pen/NEOQPQ

FIX:
https://codepen.io/drdogbot7/pen/GwYVNz

## Types of changes
Only css affecting core/columns.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

closes #13035

drdogbot7 committed:
Update style.scss

Columns were broken b/w the small and medium breakpoints, and when there are an odd number of columns. This fixes those issues.

orig: https://codepen.io/drdogbot7/pen/NEOQPQ
Fix: https://codepen.io/drdogbot7/pen/GwYVNz

drdogbot7 committed:
clarify comments

drdogbot7 committed:
apply fix in editor

drdogbot7 committed:
clarify comments

jasmussen reviewed (dismissed):
This is such a nice pull request. Meaning to address a specific problem, well documented, lots of thoughts put into it. Thanks so much. Once we get this in you should be sure to add yourself to the contributors.md doc if you'd like and if you're not already there.

How does this one relate to #12199? Should it be merged along with the other, or is one of them sufficient?

I took this for a spin and especially on the frontend this seems to work well:

![frontend](https://user-images.githubusercontent.com/1204802/50073338-9839cf80-01d8-11e9-8f4c-405b1487ad44.gif)

However there seems to be a small discrepancy in the editor view. Here's 5.0.1:

<img width="866" alt="screenshot 2018-12-17 at 08 45 42" src="https://user-images.githubusercontent.com/1204802/50073354-a851af00-01d8-11e9-8fef-4b3105daf065.png">

Here's this branch:

<img width="797" alt="screenshot 2018-12-17 at 08 46 01" src="https://user-images.githubusercontent.com/1204802/50073364-ad166300-01d8-11e9-93c3-1cdf52d3ca38.png">

It's the same in 2019:

<img width="1174" alt="screenshot 2018-12-17 at 08 42 37" src="https://user-images.githubusercontent.com/1204802/50073385-b8698e80-01d8-11e9-911f-108a8574bff1.png">

I'm not sure quite what it is, the CSS seems to be on point. Can you reproduce? I saw this with 2 columns.

I have some comments about the code comments as well, but I will leave those separately. But overall, this is a great PR and with a few small tweaks I think we should get it in!

jasmussen commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -89,13 +89,17 @@
 			// Beyond mobile, allow 2 columns.
 			@include break-small() {
-				flex-basis: 50%;
+				// Size should be half the available width, minus the margins added below
```
@tofumatt has taught me to be consistent (going forward, I was a bad actor for a long time) with our code comment system. Which is capital first letter, end with a period.

In this case we're missing a period at the end.

jasmussen commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -89,13 +89,17 @@
 			@include break-small() {
-				flex-basis: 50%;
+				// Size should be half the available width, minus the margins added below
+				// In the editor, we also need to account for additional block padding
```
Same here. I know this is really nitpicky, but we're basically a margin fix and some code comments polish from shipping this.

jasmussen commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -105,11 +109,19 @@
+				&:nth-child(odd) {
+					margin-right: initial;
+				}
+				// Add left margin to all but first column
```
Needs a period.

drdogbot7 replied:
thanks.

jasmussen commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -105,11 +109,19 @@
 					margin-left: $grid-size-large * 2;
 				}
-
+				// add right margin to all but last column
```
Capital first letter and period.

jasmussen commented on packages/block-library/src/columns/style.scss:
```diff
@@ -18,7 +18,9 @@
 	@include break-small() {
-		flex-basis: 50%;
+
+		// Size should be half the available width, minus the margins added below
```
Needs a period. Hey, actually I think I'll stop these reviews and just push a fix. Hold on.

Commit:
Polish up comments.

jasmussen commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -89,27 +89,43 @@
-				flex-basis: 50%;
+				// Size should be half the available width, minus the margins added below.
+				// In the editor, we also need to account for additional block padding.
+				flex-basis: calc(50% - #{$grid-size-large * 2} - #{$block-padding * 2});
```
I took a stab at fixing the margin issue I noted. When I reverted this back to `flex-basis: 50%;` (editor only), I got things working:

<img width="813" alt="screenshot 2018-12-17 at 09 03 05" src="https://user-images.githubusercontent.com/1204802/50073958-ab4d9f00-01da-11e9-80f2-81b2d07f3b40.png">

Keep in mind that the editor uses a lot of negative margins left and right so as to make available a selectable/hoverable space left and right of blocks. This might be why the additional grid stuff isn't necessary here. Am I missing something? 

This is with `flex-basis: 50%;`:

<img width="813" alt="screenshot 2018-12-17 at 09 03 05" src="https://user-images.githubusercontent.com/1204802/50074026-dc2dd400-01da-11e9-8acb-341f642c53fd.png">

<img width="918" alt="screenshot 2018-12-17 at 09 04 47" src="https://user-images.githubusercontent.com/1204802/50074030-df28c480-01da-11e9-89d2-3ea6c23a985d.png">

Again, I have no doubt that this is necessary on the front-end, but I think it's what's causing issues in the editor.

drdogbot7 replied:
Hm. When I change this back to `flex-basis: 50%;`, this is what I get at the Small (600px) breakpoint.

Did you have to make any other changes?

![image](https://user-images.githubusercontent.com/1050936/50093634-e057f280-01d6-11e9-83aa-c6bcb2d05a1a.png)

jasmussen replied:
Ah good call. But if you change "break-medium" back to "break-small" here: https://github.com/WordPress/gutenberg/pull/12816/files#diff-49ac41bc72e24e343a451fe254801c0fR38 — it works again:

![screenshot 2018-12-17 at 16 51 43](https://user-images.githubusercontent.com/1204802/50098373-0e135a80-021c-11e9-9798-12518e686aad.png)

drdogbot7 replied:
At the small (2-column) breakpoint… I think as long as there's a margin b/w the columns, then `flex-basis` needs to account for that doesn't it?  50% plus any margin will always wrap after the first column, right?

jasmussen replied:
Possibly. But keep in mind that there is negative margins applied to columns in the _editor only_, to ensure that this looks right. Keep in mind the markup in the editor is different from the markup on the frontend — there are a lot more divs in the editor to enable the various tools for each block. So until we have access to something like `display: contents;` there's likely always going to be a discrepancy between the frontend and backend. 

Did you try flex-basis: 50%? For me it seemed to fix the issue of the extra right margin, but there could be situations I haven't accounted for there.

jasmussen replied:
> For me flex-basis:50%; fixes the extra margin at the medium breakpoint, but only at the expense of breaking 2-column layout at the small breakpoint.

This only works if you _also_ set the wrap property to nowrap as noted in https://github.com/WordPress/gutenberg/pull/12816#discussion_r242202364.

But I'm going to take a look at your latest push, see if that fixes it!

jasmussen commented:
Pushed some polish to the comments, and left an idea for a solution in https://github.com/WordPress/gutenberg/pull/12816/files#r242051845.

drdogbot7 committed:
more consistent block padding

drdogbot7 commented:
I pushed a new commit which works for me.

In the editor I get 2 columns at the small breakpoint and no extra margins at the medium breakpoint.

I adjusted the margins so that we get the same spacing b/w columns (64px) in the editor and front-end.

I don't love using Calc b/c it feels a little dirty to me, but I don't see a better way to get 2 even columns.

Something like this might be ideal in the future, but it only works in FF currently.

```
@include break-small() {
	flex: 1;
	&::nth-child(2n) {
		page-break-after: always;
		break-after: always;
	}
}
```

One issue I still have is that the vertical spacing b/w rows is a little tight.

![image](https://user-images.githubusercontent.com/1050936/50105576-ba404b80-01f2-11e9-9ad3-96cbd515ee9b.png)

The front-end CSS adds a `margin-bottom:1em;` to each column, so maybe we ought to add that in the editor as well?

jasmussen commented:
The latest change seems to improve things a fair bit at the 2 column breakpoint.

![screenshot 2018-12-17 at 19 36 59](https://user-images.githubusercontent.com/1204802/50107645-82a5c380-0233-11e9-9963-595bb3788764.png)

![screenshot 2018-12-17 at 19 38 41](https://user-images.githubusercontent.com/1204802/50107652-85081d80-0233-11e9-8644-adecd2af5956.png)

But the margins seem too big now, compared to master. This branch:

![screenshot 2018-12-17 at 19 39 23](https://user-images.githubusercontent.com/1204802/50107656-889ba480-0233-11e9-85bd-28b23d50d520.png)

Master:

![screenshot 2018-12-17 at 19 42 16](https://user-images.githubusercontent.com/1204802/50107870-15def900-0234-11e9-9d6c-3c7ffd9c7bfe.png)

There's also an issue with many columns:
 
![screenshot 2018-12-17 at 19 40 50](https://user-images.githubusercontent.com/1204802/50107886-22fbe800-0234-11e9-8511-572666156b40.png)

Here's master:

![screenshot 2018-12-17 at 19 42 08](https://user-images.githubusercontent.com/1204802/50107898-28593280-0234-11e9-9c1a-10f700845550.png)

It's a big headache, I know. And it's getting late for me here in Denmark. I promise I'll take a good long stab at this tomorrow, look at your latest changes, and see if I can fix it up.

jasmussen commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -35,7 +35,7 @@
 		flex-wrap: wrap;
 
-		@include break-small() {
+		@include break-medium() {
```
We discussed this a little bit already, but what is the context for changing this to medium? 

To my understanding the ideal situation is this:

- Mobile (small) breakpoint has 1 column
- Medium breakpoint has 2 columns
- > Medium has user set amount of columns.

To me this says that Mobile should _wrap_, but anything above mobile should _not_ wrap.

drdogbot7 replied:
The reason to change this to medium is so that the small-to-medium breakpoint (600px - 782px) will wrap.

It is supposed to wrap _after_ the EVEN columns, so that we get 2 columns, right?

```
Mobile:
1 (wrap)
2 (wrap)
3 (wrap)
4 (wrap)
5

Small to Medium
1, 2 (wrap)
3, 4 (wrap)
5

Medium and up
1, 2, 3, 4, 5
```
If we set 'nowrap' at the Small breakpoint (600px), then we eliminate the small-to-medium breakpoint.

I'm not opposed to removing the small-to-medium breakpoint entirely; I just didn't think that was the original intent.

jasmussen replied:
That's a solid point. Thank you for clarifying, and for the patience.

Commit:
Push a few fixes:

- Reduce the margin between columns. This fixes an overlap issue at 6 columns wide.
- Improve the wrapping and mobile, by adding a bottom margin to the column container.

jasmussen commented:
Okay, I think the branch is now in a good state. I pushed a few small tweaks.

I discovered at the wide alignments there was no overlap as previously noted [here](https://github.com/WordPress/gutenberg/pull/12816#issuecomment-447954089):

<img width="1104" alt="screenshot 2018-12-18 at 09 17 51" src="https://user-images.githubusercontent.com/1204802/50141728-86bff880-02a8-11e9-8417-7d89636452aa.png">

But it was present when not wide:

<img width="777" alt="screenshot 2018-12-18 at 09 18 04" src="https://user-images.githubusercontent.com/1204802/50141738-8cb5d980-02a8-11e9-9858-9d215877b16a.png">

This suggested that the problem was lack of room, more than any other thing. So I reduced the margins, which I have come around to, were too large regardless. This is a big PR, so I think it's good to make this change along with it, especially as it solves a problem.

There was an overlap issue at the 2 column breakpoint:

<img width="661" alt="screenshot 2018-12-18 at 09 19 00" src="https://user-images.githubusercontent.com/1204802/50141788-afe08900-02a8-11e9-8a56-be2262a482fa.png">
<img width="496" alt="screenshot 2018-12-18 at 09 19 10" src="https://user-images.githubusercontent.com/1204802/50141794-b242e300-02a8-11e9-9af5-17cccb751955.png">

This was because flex causes a new margin context, which means the block margins collapse. In other words, because we wrap the column blocks (singular), _those_ now need a bottom margin. I added that:

<img width="446" alt="screenshot 2018-12-18 at 09 24 42" src="https://user-images.githubusercontent.com/1204802/50141844-d4d4fc00-02a8-11e9-9f71-994739dc7200.png">

This was also necessary for mobile:

<img width="568" alt="screenshot 2018-12-18 at 09 34 44" src="https://user-images.githubusercontent.com/1204802/50141862-e0282780-02a8-11e9-98da-42c9dd0f16b7.png">

This is now the frontend:

<img width="840" alt="screenshot 2018-12-18 at 09 35 26" src="https://user-images.githubusercontent.com/1204802/50141876-e7e7cc00-02a8-11e9-8232-cd0392d07a9f.png">

And this is the backend:

<img width="836" alt="screenshot 2018-12-18 at 09 35 31" src="https://user-images.githubusercontent.com/1204802/50141879-eae2bc80-02a8-11e9-9209-6c355613dcaf.png">

I think this is ready for a wider code review and good testing. So I'm expanding the range.

drdogbot7 reviewed (commented):
Working for me. Looks consistent between FE and Editor. 👍

m-e-h commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -89,29 +93,49 @@
+			// Beyond the medium breakpoint, show the user-set amount of columns.
+			@include break-medium() {
+				// Because columns do not wrap, reset the column margin.
+				margin-bottom: initial;
```
Looks like postcss or something else is actually converting it to `0` in the final CSS anyway. Still probably a good idea to avoid the abstraction here.

drdogbot7 replied:
> Still probably a good idea to avoid the abstraction here.

Makes sense to me. I guess using `initial` doesn't really _get us_ anything.

m-e-h commented on packages/block-library/src/columns/style.scss:
```diff
@@ -18,7 +18,9 @@
 
```
Should we change the shorthand `flex: 1` to just the property we're using from it, `flex-grow: 1`, here?

A little redundant. This is essentially what we currently have:
```
.wp-block-column {
    flex: 1;
       (flex-grow: 1)
       (flex-basis: 0%)

    flex-basis: 100%;
}
```

drdogbot7 replied:
Good call there I'd say. I still always have to lookup what `Flex:1;`actually means.

drdogbot7 replied:
Actually, I don't think we need to set `flex:1;` or `flex-grow:1;` at all.  `flex-basis:100%;` (or `width:100%;`) is all we need.

m-e-h commented on packages/block-library/src/columns/style.scss:
```diff
@@ -31,20 +33,37 @@
+	}
+
+	// All columns fit in one row at this breakpoint.
+	@include break-medium() {
```
This `even, child, nth, first, not` reset stuff makes me dizzy :dizzy_face:

Any interest in simplifying this?

Something along these lines maybe:
```scss
	@include break-medium() {

		// Fill in the left and right margin holes left by the previous small breakpoint.
		// First column already has a margin-right and no margin-left.
		&:not(:first-child) {
			margin-left: $grid-size-large;
			margin-right: $grid-size-large;
		}

		// Remove right margin on last column.
		&:last-child {
			margin-right: $grid-size-large;
		}
	}
```

m-e-h commented:
There's no real benefit to this, honestly. 

I may have gotten a little carried away trying make it simple. And it still looks complicated. 😀

jasmussen commented:
I love the great feedback on this thread. In _this_ particular case, I applaud the simplification.

It's worth noting that the smaller the change we make, the more likely it is to get in soon. Already now this is sort of a big change, though we can mitigate that with thorough testing:

- Mobile, medium, large breakpoints
- Frontend and backend both
- Physical iphones/androids (or the emulators/simulators)
- Modern browsers
- IE11 and Edge

A note on IE11 — I recall columns working there, but the experience doesn't have to be top tier, it just has to not be broken.

drdogbot7 commented:
>  In this particular case, I applaud the simplification.

If we want to simplify and make it easy to understand, **_and_** we're willing to make significant changes then:

* **Remove the `:not` rules.**  I like how clever they are, but they confused me at first, and I know I'm not the only one.

* **Remove the Small-to-medium breakpoint.** That seems to be the source of a lot of the complexity and the confusion. Removing it will make things easier on themers who want to customize.  It currently doesn't work, so removing it shouldn't inconvenience too many people.

Now here's our entire style.scss
```
.wp-block-columns {
	display: flex;

	// Responsiveness: Allow wrapping on mobile.
	flex-wrap: wrap;

	@include break-medium() {
		flex-wrap: nowrap;
	}
}

.wp-block-column {
	width: 100%; // works just like flex-basis here, and it's easier to override
	margin-bottom: 1em;

	// Prevent the columns from growing wider than their distributed sizes.
	min-width: 0;

	// Prevent long unbroken words from overflowing.
	word-break: break-word; // For back-compat.
	overflow-wrap: break-word; // New standard.

	// User defined # of columns beyond medium breakpoint.
	@include break-medium() {
		margin-left: $grid-size-large;
		margin-right: $grid-size-large;

		// Remove outside margin from first and last
		&:first-child {
			margin-left: 0;
		}

		&:last-child {
			margin-right: 0;
		}
	}
}
```

> It's worth noting that the smaller the change we make, the more likely it is to get in soon.

This was my thinking originally, but if we're going to really simplify, then let's do it.

m-e-h commented:
That's similar to how the text-columns are doing it. https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/text-columns/style.scss

Skinning this cat could go on and on. Since the columns are actually "broken" currently, we should probably get a fix in sooner than later.

Is there still a text-column block, btw? Not sure I've ever seen it...

jasmussen commented:
So, that brings us to the big question. When can we ship this?

We have 5.0.x releases, which are supposed to be really small or really critical. And we have major releases which can be bigger, like 5.1. The latter obviously is a while out. 

@youknowriad if you have time — what are the chances for this pull request to make it to a minor release? And if it's too big as is, how small would it have to be to land in a minor release?

youknowriad commented:
I think this can go in 5.1 which is targeted for February. At this point it's not certain whether there will be a 5.0.3 before that.

drdogbot7 committed:
set margins to 0 not initial, rm unnecessary flex properties, change flex-basis to width

drdogbot7 commented:
I pushed a small update.

- Changed `margin-*: initial;` to `margin-*: 0;`, as per @m-e-h's comment.
- Removed unnecessary `flex: 1;` and `flex-grow: 0;`
- Changed `flex-basis` to `width`. It does the same thing, is slightly more compatible (IE), and is easier to override. (content < width < flex-basis).

I did basic testing of editor and FE in Chrome, Safari, Firefox (all mac), as well as IE11.

mapk reviewed (commented):
Looking really good!

I tested this with a few different themes. Take a look at these gifs. The first two themes still showed an indention issue, but the last theme, Twenty Nineteen, appeared fine.

https://cloudup.com/cl64LZYWXyZ

I tested on MacBook Pro, Chrome 71.0.

I'm not entirely sure if this is a theme issue, or something else. If I can help with providing more info, please let me know.

WraithKenny commented:
This is all needlessly complex and with too many rules that themes have to overwrite to fix.

All columns should have both left and right spacing. There should be no "odd" or "first/last" checks. The wrapping element should have negative margins to nullify those margins.

Please borrow the knowledge of the well-worn cowpaths trodden by Bootstrap, and don't rediscover and repeat the same mistakes from decades past. https://getbootstrap.com/docs/4.1/layout/grid/

m-e-h commented:
I mostly agree @WraithKenny . I at one point had a version which eliminated the even, odd stuff.
The problem with the negative margins on the wrapper though, is that a common pattern I'm seeing themes use for blocks is something similar to this:
```
.entry-content > * {
    max-width: 900px;
    margin-left: auto;
    margin-right: auto;
}
```
In this case, a negative left margin would take this block out of the center and move it to the left of the screen.

Like I said though, I do agree. There have got to be simpler methods.

WraithKenny commented:
@m-e-h Yeah, 2019 uses something of the sort, which I assume you are referring to. I think the effect you are talking about is **less** broken looking than exists currently on all sites that don't have special css overwrites for this. And 2019 can be patched and released with 5.1.

drdogbot7 commented:
> This is all needlessly complex and with too many rules that themes have to overwrite to fix.

I do agree with @WraithKenny and @m-e-h that simpler would be better for the sake of theming, but the point of _this_ PR is just to _fix the bugs_ that were present in the initial release, not to make any other changes.

drdogbot7 commented:
> Looking really good!
> 
> I tested this with a few different themes. Take a look at these gifs. The first two themes still showed an indention issue, but the last theme, Twenty Nineteen, appeared fine.
> 
> https://cloudup.com/cl64LZYWXyZ
> 
> I tested on MacBook Pro, Chrome 71.0.
> 
> I'm not entirely sure if this is a theme issue, or something else. If I can help with providing more info, please let me know.

I just tested in Atomic Theme and Monostack and I couldn't replicate these issues that @mapk had. Possibly themes have been updated?

Atomic Theme
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/1050936/51286011-c288d380-19b6-11e9-8287-7516dabe823a.gif)

Monostack:
![ezgif com-video-to-gif 1](https://user-images.githubusercontent.com/1050936/51286018-c9174b00-19b6-11e9-843e-03f696e9b155.gif)

drdogbot7 commented:
Closing this to avoid confusion. #12199 has similar fixes.
